pkg/loop/config.go

```diff
@@ -0,0 +1,224 @@
+package loop
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/hashicorp/go-plugin"
+	libocr "github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	pb "github.com/smartcontractkit/chainlink-relay/pkg/loop/internal/pb"
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+)
+
+var _ types.ConfigProvider = (*configProviderClient)(nil)
+
+type configProviderClient struct {
+	providerClient
+	offchainDigester libocr.OffchainConfigDigester
+	contractTracker  libocr.ContractConfigTracker
+}
+
+func newConfigProviderClient(id uint32, broker *plugin.GRPCBroker) *configProviderClient {
+	return &configProviderClient{providerClient: providerClient{id: id, broker: broker}}
+}
+
+func (c *configProviderClient) Start(ctx context.Context) error {
+	if err := c.providerClient.Start(ctx); err != nil {
+		return err
+	}
+	c.offchainDigester = &offchainConfigDigesterClient{pb.NewOffchainConfigDigesterClient(c.cc)}
+	c.contractTracker = &contractConfigTrackerClient{pb.NewContractConfigTrackerClient(c.cc)}
+	return nil
+}
+
+func (c *configProviderClient) OffchainConfigDigester() libocr.OffchainConfigDigester {
+	return c.offchainDigester
+}
+
+func (c *configProviderClient) ContractConfigTracker() libocr.ContractConfigTracker {
+	return c.contractTracker
+}
+
+var _ libocr.OffchainConfigDigester = (*offchainConfigDigesterClient)(nil)
+
+type offchainConfigDigesterClient struct {
+	grpc pb.OffchainConfigDigesterClient
+}
+
+func (o *offchainConfigDigesterClient) ConfigDigest(config libocr.ContractConfig) (digest libocr.ConfigDigest, err error) {
+	pbCfg := pbContractConfig(config)
+	var reply *pb.ConfigDigestReply
+	reply, err = o.grpc.ConfigDigest(context.TODO(), &pb.ConfigDigestRequest{
+		ContractConfig: pbCfg,
+	})
+	if err != nil {
+		return
+	}
+	if l := len(reply.ConfigDigest); l != 32 {
+		err = fmt.Errorf("wrong digest length: expect 32 but got %d", l) //TODO necessary? then add more places?
+		return
+	}
+	copy(digest[:], reply.ConfigDigest)
+	return
+}
+
+func (o *offchainConfigDigesterClient) ConfigDigestPrefix() libocr.ConfigDigestPrefix {
+	reply, err := o.grpc.ConfigDigestPrefix(context.TODO(), &pb.ConfigDigestPrefixRequest{})
```

Comment:
What did you have in mind? I don't think there are any real workarounds unfortunately. Context is not such a big loss, but without anywhere to send the errors we either have to panic or return a phony value. I'm hoping to follow up with https://smartcontract-it.atlassian.net/browse/BCF-2073 right away (or even in parallel if we want to hold on merging this :shrug:).

Reply:
i was just thinking you could handle a context with a `select` and return an error if context was cancelled early

and then for panics - i vaguely recall panic handling to be a thing but not sure how it technically works? if we can "catch" a panic - then we can return the panic error upstream

Reply:
> i was just thinking you could handle a context with a select and return an error if context was cancelled early

Oh yeah that would be possible, although I'm not sure that those cases actually block, since the client side grpc caller could just move on after timeout anyways.

Yeah we could try to panic and recover, but the stack goes through libocr in complicated ways (e.g. `ReportPlugin`s are highly concurrent), so it may not be possible to do without refactoring those internals is a way that is even more disruptive than regular err/ctx.